sdk/core/src/bucket.rs

```diff
@@ -0,0 +1,94 @@
+use ic_kit::ic::call;
+use ic_kit::{Principal, RejectionCode};
+
+use crate::root::RootBucket;
+use ic_history_common::{
+    GetIndexCanistersResponse, GetTransactionResponse, GetTransactionsArg, GetTransactionsResponse,
+    GetUserTransactionsArg, WithIdArg, WithWitnessArg,
+};
+
+/// A contract-specific bucket canister.
+///
+/// A bucket canister implements storage for its parent contract. The total storage for a given
+/// contract is created using multiple bucket canisters, which are interconnected using a root bucket
+/// and router system. Querying buckets also features pagination.
+#[derive(Copy, Clone)]
+pub struct Bucket(pub(crate) Principal);
+
+impl Bucket {
+    /// Returns the list of canisters which have different pages of data.
+    pub async fn get_next_canisters(&self) -> Result<Vec<Bucket>, (RejectionCode, String)> {
+        let result: (GetIndexCanistersResponse,) = call(
+            self.0,
+            "get_next_canisters",
+            (WithWitnessArg { witness: false },),
+        )
+        .await?;
+
+        Ok(result
+            .0
+            .canisters
+            .iter()
+            .map(|canister| Bucket(*canister))
+            .collect())
```

Comment:
I really wonder if the compiler can  actually convert this code to no-op since Bucket has essentially the same layout as a Principal and we're doing a `Vec<Principal>` to `Vec<Bucket>`, so in theory it's the same as a transmute

Reply:
Interesting thought, since everything is `Copy` at the very least it should be speedy as heck

Reply:
https://godbolt.org/z/jh16bG7Y4